modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4IncrementalRequestHandlingIT.java

```diff
@@ -0,0 +1,429 @@
+        try (var ctx = setupClientCtx()) {
+            var opaqueId = opaqueId(0);
+            var payloadSize = MBytes(50);
+            ctx.clientChannel.writeAndFlush(httpRequest(opaqueId, payloadSize));
+            for (int i = 0; i < 5; i++) {
+                ctx.clientChannel.writeAndFlush(randomContent(MBytes(10), false));
+            }
+            ctx.clientChannel.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
+
+            var handler = ctx.awaitRestChannelAccepted(opaqueId);
+
+            // Read buffers for socket and channel usually within few MBytes range all together.
+            // This test assumes that buffers will not exceed 10 MBytes, in other words there should
+            // be less than 10 MBytes in fly between http client's socket and rest handler. This
+            // loop ensures that reading 10 MBytes of content on server side should free almost
+            // same size in client's channel write buffer.
+            for (int mb = 0; mb <= 50; mb += 10) {
+                var minBufSize = payloadSize - MBytes(10 + mb);
+                var maxBufSize = payloadSize - MBytes(mb);
+                assertBusy(() -> {
+                    var bufSize = ctx.clientChannel.bytesBeforeWritable();
+                    assertTrue(
+                        "client's channel buffer should be in range [" + minBufSize + "," + maxBufSize + "], got " + bufSize,
+                        bufSize >= minBufSize && bufSize <= maxBufSize
+                    );
+                });
+                handler.consumeBytes(MBytes(10));
+            }
+            assertTrue(handler.stream.hasLast());
+        }
+    }
+
+    private String opaqueId(int reqNo) {
+        return getTestName() + "-" + reqNo;
+    }
+
+    static int MBytes(int m) {
+        return m * 1024 * 1024;
+    }
+
+    static <T> T safePoll(BlockingDeque<T> queue) {
+        try {
+            var t = queue.poll(SAFE_AWAIT_TIMEOUT.seconds(), TimeUnit.SECONDS);
+            assertNotNull("queue is empty", t);
+            return t;
+        } catch (InterruptedException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    static FullHttpRequest fullHttpRequest(String opaqueId, ByteBuf content) {
+        var req = new DefaultFullHttpRequest(HTTP_1_1, POST, ControlServerRequestPlugin.ROUTE, Unpooled.wrappedBuffer(content));
+        req.headers().add(CONTENT_LENGTH, content.readableBytes());
+        req.headers().add(CONTENT_TYPE, APPLICATION_JSON);
+        req.headers().add(Task.X_OPAQUE_ID_HTTP_HEADER, opaqueId);
+        return req;
+    }
+
+    static HttpRequest httpRequest(String opaqueId, int contentLength) {
+        return httpRequest(ControlServerRequestPlugin.ROUTE, opaqueId, contentLength);
+    }
+
+    static HttpRequest httpRequest(String uri, String opaqueId, int contentLength) {
+        var req = new DefaultHttpRequest(HTTP_1_1, POST, uri);
+        req.headers().add(CONTENT_LENGTH, contentLength);
```

Comment:
Could we also have some chunked requests which don't specify the content-length up front?